web_poet/overrides.py

```diff
@@ -0,0 +1,266 @@
+        :param overrides: The Page Object that should be replaced by the annotated one.
+        :param exclude: Defines URLs over which the override should not happen.
+        :param priority: The resolution priority in case of conflicting annotations.
+        """
+
+        def wrapper(cls):
+            self._declare_registry_in_module(cls)
+
+            spec = HandleUrlsSpec(
+                patterns=Patterns(
+                    include=_as_list(include),
+                    exclude=_as_list(exclude),
+                    priority=priority,
+                ),
+                overrides=overrides,
+                meta=kwargs,
+            )
+            # If it was already defined, we don't want to override it
+            if cls not in self.data:
+                self.data[cls] = spec
+            else:
+                warnings.warn(
+                    f"Multiple @handle_urls annotations with the same 'overrides' "
+                    f"are ignored in the same Registry. Ignoring duplicate "
+                    f"annotation on '{include}' for {cls}."
+                )
+
+            return cls
+
+        return wrapper
+
+    def get_data_from_module(self, module: str) -> Dict[Callable, HandleUrlsSpec]:
+        """Returns the override mappings that were declared using ``handle_urls``
+        in a specific module.
+
+        This is useful if you've organized your Page Objects into multiple
+        submodules in your project.
+        """
+        return {
+            cls: spec
+            for cls, spec in self.data.items()
+            if cls.__module__.startswith(module)
+        }
+
+    def __repr__(self) -> str:
+        return f"PageObjectRegistry(name='{self.name}')"
+
+
+# For ease of use, we'll create a default registry so that users can simply
+# use its `handles_url()` method directly by `from web_poet import handles_url`
+default_registry = PageObjectRegistry(name="default")
+handle_urls = default_registry.handle_urls
+
+
+def walk_modules(module: str) -> Iterable:
+    """
+    Return all modules from a module recursively. Note that this will import all the modules and submodules.
+    It returns the provided module as well.
+    """
+
+    def onerror(err):
+        raise err  # pragma: no cover
+
+    spec = importlib.util.find_spec(module)
+    if not spec:
+        raise ImportError(f"Module {module} not found")
+    mod = importlib.import_module(spec.name)
+    yield mod
+    if spec.submodule_search_locations:
+        for info in pkgutil.walk_packages(
+            spec.submodule_search_locations, f"{spec.name}.", onerror
+        ):
+            mod = importlib.import_module(info.name)
+            yield mod
+
+
+def find_page_object_overrides(
+    module: str, registry_name: str = "default"
+) -> List[OverrideRule]:
+    """
+    Find all the Page Objects overrides in the given module/package and its
+    submodules.
```

Comment:
@kmike It's primarily used by the upcoming revised Registry format from this `scrapy-poet`'s PR: https://github.com/scrapinghub/scrapy-poet/pull/56/files. It's main use case is to read all `@handle_urls` annotation to populate this setting:

```
from web_poet import find_page_object_overrides

SCRAPY_POET_OVERRIDES = find_page_object_overrides("my_scrapy_project.page_objects")
```

It's quite convenient since the user doesn't need to interact directly with `web-poet`'s own `PageObjectRegistry` class. Moreover, should we decide to update the implementation away from `PageObjectRegistry` class, we could prevent any backward incompatible changes using this interface. _(We could then deprecate `PageObjectRegistry` class slowly but still present for backward compatibility)_

Reply:
I see, thanks! I wonder how it'd look like if we remove the indirection.

```
from web_poet import default_registry
walk_modules("my_scrapy_project.page_objects")
SCRAPY_POET_OVERRIDES = default_registry.get_overrides()

# we can expose it as a function as well, e.g. 
# SCRAPY_POET_OVERRIDES = web_poet.get_overrides()
```
If you do this change, it'd allow to remove:

* name property from PageObjectRegistry
* `_declare_registry_in_module` method, as well as magic REGISTRY_MODULE_ANCHOR cache
* get_data_from_module mehod
* find_page_object_overrides funcion
* probably, it'd be possible to drop HandleUrlsSpec, and just use OverrideRule, not sure

The way it could work:

1. We should make sure walk_modules is called.
2. All work is done on import time, in handle_urls decorators; all results are stored in the registry object.

For me it looks like a big simplification of the system, but I may be missing some features or use cases.

Reply:
I see your point. 👍 Yeah, I think it make sense to remove `find_page_object_overrides()` if we'd directly expose the registry here. It would seem that the new `PageObjectRegistry` completely removes the need for such function since we're not dealing with the _magic_ module attributes anymore from the previous refactor.

I'd propose to have the following interfaces as I think a separate call to `walk_modules()` outside of the registry would introduce some complexity.

```python
from web_poet import default_registry

# retrieves all overrides inside the current Scrapy project
SCRAPY_POET_OVERRIDES = default_registry.get_overrides()

# or, could retrieve them from specific modules 
SCRAPY_POET_OVERRIDES = default_registry.get_overrides_from_module("common_page_objects.ecommerce.lalazon")
```

I'll try to do a soft refactor and see what the new code would look like. But I really do think this would simplify it a lot. Thanks for bringing it up!

Reply:
Walking the current Scrapy project makes sene.

I'm not sure about `default_registry.get_overrides_from_module("common_page_objects.ecommerce.lalazon")`.
walk_modules makes sure that all modules in a package are imported, and so decorators are executed.
But there can be modules outside, which are also imported. So, if get_overrides_from_module is implemented as "walk_modules; return self.overrides", it may return overrides which are not in `"common_page_objects.ecommerce.lalazon"`.

Then, the next question: do we actually want to have such feature, "return only POs from this package"? For me it looks we're fine without this feature; users can use separate registries if they want.

Reply:
> Then, the next question: do we actually want to have such feature, "return only POs from this package"? For me it looks we're fine without this feature; users can use separate registries if they want.

During the middle of the refactoring process on this, I realized that collecting the overrides rules using a module string works really well when users organize and sort their Page Objects into directories. We've tackled such PO organization before with something like:

```
my_scrapy_project
|
└── page_objects
    ├── lalazon
    |   ├── base
    |   |   ├── __init__.py
    |   |   ├── lalazon_products.py
    |   |   └── lalazon_product_listings.py
    |   ├── us
    |   |   ├── __init__.py
    |   |   ├── lalazon_us_products.py
    |   |   └── lalazon_us_product_listings.py
    |   └── fr
    |       ├── __init__.py
    |       ├── lalazon_fr_products.py
    |       └── lalazon_fr_product_listings.py
    └── rawrmart
        ├── __init__.py
        ├── rawrmart_products.py
        └── rawrmart_product_listings.py
```

Given the Scrapy project structure above, users could opt with just a single registry which is `default_registry`. They would then have the capability to filter the rules per module.

```python
from web_poet import default_registry

SCRAPY_POET_OVERRIDES = default_registry.get_overrides_from_module("my_scrapy_project.page_objects.lalazon")
# or 
SCRAPY_POET_OVERRIDES = default_registry.get_overrides_from_module("my_scrapy_project.page_objects.rawrmart")
```

Here's a WIP commit for the ongoing refactor in case your interested on how it turned out: https://github.com/scrapinghub/web-poet/pull/16/commits/1efb5068ed1642bc04c8dd0ba814377a910eb28d. I've yet to update the docs, implement some methods, and add some more tests.

Reply:
Updated the docs for https://github.com/scrapinghub/scrapy-poet/pull/56 in-line with this recent refactoring: https://github.com/scrapinghub/scrapy-poet/pull/56/commits/c865c60f1cea0fba17186fd230391be0b0bb1577

Reply:
@kmike Our last offline conversation ended on the thought of simplifying how the Override Rules are declared and organized by choosing either the **package-based** or **Multiple Registry** approach. However, after giving some time to think about it, I still believe it's worth keeping both approaches present to give adequate amount of options to developers.

I wrote down a dedicated section on the topic in our docs using this commit: https://github.com/scrapinghub/web-poet/pull/16/commits/75593edb29378299093dc80ae97c9e3fa483ee13. I believe this could clarify the advantages of keeping both approaches available by exploring different scenarios in using them.

Reply:
Resolving this thread as we've branched out into newer discussions in separate threads.